File: backend/app/models/domain/cluster.py
"""Cluster domain models."""

from datetime import datetime

from pydantic import BaseModel, Field

from app.models.enums.common import ClusterStatus, ComplaintCategory
from app.models.schemas.common import DocumentMetadata, DocumentMetadataCreate, GeoLocation


class ClusterAIInsights(BaseModel):
    """AI-generated insights for a complaint cluster."""

    summary: str = Field(max_length=3000)
    root_causes: list[str] = Field(default_factory=list, max_length=20)
    affected_population_estimate: int | None = Field(default=None, ge=0)
    priority_score: float = Field(ge=0, le=1)
    key_themes: list[str] = Field(default_factory=list, max_length=20)
    generated_at: datetime
    model_version: str = Field(default="gemini-2.0-flash", max_length=64)


class ClusterPriorityAnalysis(BaseModel):
    """Gemini-powered government priority intelligence for a cluster."""

    priority_score: int = Field(ge=0, le=100)
    impact_score: int = Field(ge=0, le=100)
    urgency_level: str = Field(max_length=32)
    risk_level: str = Field(max_length=32)
    affected_population_estimate: str = Field(max_length=512)
    public_safety_risk: str = Field(max_length=2000)
    infrastructure_criticality: str = Field(max_length=2000)
    environmental_impact: str = Field(max_length=2000)
    economic_impact: str = Field(max_length=2000)
    suggested_department: str = Field(max_length=256)
    recommended_action: str = Field(max_length=3000)
    estimated_resolution_time: str = Field(max_length=256)
    estimated_budget_range: str = Field(max_length=256)
    priority_rank: int | None = Field(default=None, ge=1)
    reasoning: str = Field(max_length=5000)
    confidence_score: float = Field(ge=0, le=1)
    contributing_factors: list[str] = Field(default_factory=list, max_length=20)
    expected_impact: str = Field(max_length=2000)
    estimated_beneficiaries: str = Field(max_length=512)
    why_priority_ranked_high: str = Field(max_length=2000)
    analysis_hash: str = Field(max_length=64)
    processed_at: datetime
    model_version: str = Field(default="gemini-2.5-flash", max_length=64)
    prompt_version: str = Field(default="1.0.0", max_length=16)


class ClusterBase(BaseModel):
    """Shared cluster fields."""

    title: str = Field(min_length=5, max_length=256)
    description: str = Field(min_length=10, max_length=5000)
    theme: str = Field(min_length=3, max_length=256)
    category: ComplaintCategory
    status: ClusterStatus = ClusterStatus.OPEN
    complaint_ids: list[str] = Field(default_factory=list, max_length=500)
    complaint_refs: list[str] = Field(
        default_factory=list,
        max_length=500,
        description="Firestore paths: complaints/{id}",
    )
    complaint_count: int = Field(default=0, ge=0)
    village_ids: list[str] = Field(default_factory=list, max_length=200)
    village_names: list[str] = Field(default_factory=list, max_length=200)
    village_name: str | None = Field(default=None, max_length=128)
    department: str | None = Field(default=None, max_length=256)
    coordinates: GeoLocation | None = None
    representative_complaint_id: str | None = Field(default=None, max_length=128)
    average_severity: str | None = Field(default=None, max_length=32)
    latest_complaint_date: datetime | None = None
    average_confidence: float | None = Field(default=None, ge=0, le=1)
    affected_area: str | None = Field(default=None, max_length=512)
    priority_score: float = Field(default=0.5, ge=0, le=1)
    hotspot_score: float = Field(default=0.0, ge=0, le=1)
    constituency: str = Field(max_length=128)
    district: str = Field(max_length=128)
    state: str = Field(max_length=64)
    ai_insights: ClusterAIInsights | None = None
    priority_analysis: ClusterPriorityAnalysis | None = None
    impact_score: int | None = Field(default=None, ge=0, le=100)
    urgency_level: str | None = Field(default=None, max_length=32)
    priority_rank: int | None = Field(default=None, ge=1)
    recommended_department: str | None = Field(default=None, max_length=256)
    recommended_action: str | None = Field(default=None, max_length=3000)
    estimated_budget: str | None = Field(default=None, max_length=256)
    estimated_resolution_time: str | None = Field(default=None, max_length=256)
    affected_population: str | None = Field(default=None, max_length=512)
    priority_reasoning: str | None = Field(default=None, max_length=5000)
    priority_confidence: float | None = Field(default=None, ge=0, le=1)
    priority_updated_at: datetime | None = None
    priority_analysis_hash: str | None = Field(default=None, max_length=64)


class ClusterCreate(BaseModel):
    """Fields required to create a cluster."""

    title: str = Field(min_length=5, max_length=256)
    description: str = Field(min_length=10, max_length=5000)
    theme: str = Field(min_length=3, max_length=256)
    category: ComplaintCategory
    complaint_ids: list[str] = Field(default_factory=list, max_length=500)
    village_ids: list[str] = Field(default_factory=list, max_length=200)
    constituency: str = Field(max_length=128)
    district: str = Field(max_length=128)
    state: str = Field(max_length=64)
    metadata: DocumentMetadataCreate = Field(default_factory=DocumentMetadataCreate)


class ClusterUpdate(BaseModel):
    """Fields allowed on cluster update."""

    title: str | None = Field(default=None, min_length=5, max_length=256)
    description: str | None = Field(default=None, min_length=10, max_length=5000)
    theme: str | None = Field(default=None, min_length=3, max_length=256)
    category: ComplaintCategory | None = None
    status: ClusterStatus | None = None
    complaint_ids: list[str] | None = None
    complaint_refs: list[str] | None = None
    complaint_count: int | None = Field(default=None, ge=0)
    village_ids: list[str] | None = None
    village_names: list[str] | None = None
    village_name: str | None = Field(default=None, max_length=128)
    department: str | None = Field(default=None, max_length=256)
    coordinates: GeoLocation | None = None
    representative_complaint_id: str | None = Field(default=None, max_length=128)
    average_severity: str | None = Field(default=None, max_length=32)
    latest_complaint_date: datetime | None = None
    average_confidence: float | None = Field(default=None, ge=0, le=1)
    affected_area: str | None = Field(default=None, max_length=512)
    priority_score: float | None = Field(default=None, ge=0, le=1)
    hotspot_score: float | None = Field(default=None, ge=0, le=1)
    priority_analysis: ClusterPriorityAnalysis | None = None
    impact_score: int | None = Field(default=None, ge=0, le=100)
    urgency_level: str | None = Field(default=None, max_length=32)
    priority_rank: int | None = Field(default=None, ge=1)
    recommended_department: str | None = Field(default=None, max_length=256)
    recommended_action: str | None = Field(default=None, max_length=3000)
    estimated_budget: str | None = Field(default=None, max_length=256)
    estimated_resolution_time: str | None = Field(default=None, max_length=256)
    affected_population: str | None = Field(default=None, max_length=512)
    priority_reasoning: str | None = None
    priority_confidence: float | None = Field(default=None, ge=0, le=1)
    priority_updated_at: datetime | None = None
    priority_analysis_hash: str | None = Field(default=None, max_length=64)
    ai_insights: ClusterAIInsights | None = None
    updated_by: str = Field(default="system", max_length=128)


class ClusterResponse(ClusterBase):
    """Cluster document returned from Firestore."""

    id: str
    metadata: DocumentMetadata
    recommendation_ids: list[str] = Field(default_factory=list, max_length=100)


class ClusterSearchFilters(BaseModel):
    """Filter parameters for cluster queries."""

    constituency: str | None = None
    district: str | None = None
    state: str | None = None
    village_id: str | None = None
    category: ComplaintCategory | None = None
    status: ClusterStatus | None = None
    min_complaint_count: int | None = Field(default=None, ge=0)
    theme_prefix: str | None = Field(default=None, max_length=256)

File: frontend/src/types/cluster.ts
export type ClusterStatus =
  | "open"
  | "analyzing"
  | "recommended"
  | "assigned"
  | "resolved"
  | "archived";

export interface ClusterListItem {
  id: string;
  title: string;
  theme: string;
  category: string;
  status: ClusterStatus;
  department?: string | null;
  village_name?: string | null;
  complaint_count: number;
  average_severity?: string | null;
  latest_complaint_date?: string | null;
  representative_complaint_id?: string | null;
  priority_score: number;
  hotspot_score: number;
  average_confidence?: number | null;
  affected_area?: string | null;
  impact_score?: number | null;
  urgency_level?: string | null;
  priority_rank?: number | null;
  recommended_department?: string | null;
  recommended_action?: string | null;
  priority_confidence?: number | null;
  priority_updated_at?: string | null;
}

export interface ClusterComplaintSummary {
  id: string;
  title: string;
  status: string;
  village_name: string;
  submitted_at: string;
  is_duplicate: boolean;
  duplicate_score?: number | null;
  priority: string;
}

export interface ClusterPriorityAnalysis {
  priority_score: number;
  impact_score: number;
  urgency_level: string;
  risk_level: string;
  affected_population_estimate: string;
  public_safety_risk: string;
  infrastructure_criticality: string;
  environmental_impact: string;
  economic_impact: string;
  suggested_department: string;
  recommended_action: string;
  estimated_resolution_time: string;
  estimated_budget_range: string;
  priority_rank?: number | null;
  reasoning: string;
  confidence_score: number;
  contributing_factors: string[];
  expected_impact: string;
  estimated_beneficiaries: string;
  why_priority_ranked_high: string;
  analysis_hash: string;
  processed_at: string;
  model_version: string;
  prompt_version: string;
}

export interface ClusterDetail {
  id: string;
  title: string;
  description: string;
  theme: string;
  category: string;
  status: ClusterStatus;
  department?: string | null;
  village_name?: string | null;
  village_names: string[];
  coordinates?: {
    latitude: number;
    longitude: number;
    address?: string | null;
  } | null;
  complaint_ids: string[];
  complaint_count: number;
  representative_complaint_id?: string | null;
  average_severity?: string | null;
  latest_complaint_date?: string | null;
  average_confidence?: number | null;
  affected_area?: string | null;
  priority_score: number;
  hotspot_score: number;
  constituency: string;
  district: string;
  state: string;
  priority_analysis?: ClusterPriorityAnalysis | null;
  impact_score?: number | null;
  urgency_level?: string | null;
  priority_rank?: number | null;
  recommended_department?: string | null;
  recommended_action?: string | null;
  estimated_budget?: string | null;
  estimated_resolution_time?: string | null;
  affected_population?: string | null;
  priority_reasoning?: string | null;
  priority_confidence?: number | null;
  priority_updated_at?: string | null;
  related_complaints: ClusterComplaintSummary[];
}

export interface ClusterListResponse {
  success: boolean;
  items: ClusterListItem[];
  total: number;
  page: number;
  page_size: number;
  has_more: boolean;
}

export interface ClusterDashboardSummary {
  success: boolean;
  total_clusters: number;
  open_clusters: number;
  total_clustered_complaints: number;
  average_cluster_size: number;
  top_hotspots: ClusterListItem[];
}

export interface ClusterProcessResponse {
  success: boolean;
  message: string;
  complaint_id: string;
  cluster_id?: string | null;
  is_duplicate: boolean;
  duplicate_score?: number | null;
}
